src/pid-file/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+
+#![warn(missing_docs)]
+
+use std::convert::TryInto;
+use std::ffi::{CString, NulError};
+use std::fmt;
+use std::fs::Permissions;
+use std::io;
+use std::os::unix::ffi::OsStrExt;
+use std::os::unix::fs::PermissionsExt;
+use std::path::Path;
+use std::ptr;
+
+use ore::option::OptionExt;
+
+#[allow(non_camel_case_types)]
+#[repr(C)]
+struct pidfh {
+    private: [u8; 0],
+}
+
+extern "C" {
+    fn pidfile_open(
+        path: *const libc::c_char,
+        mode: libc::mode_t,
+        pid: *const libc::pid_t,
+    ) -> *mut pidfh;
+    fn pidfile_write(pfh: *mut pidfh) -> libc::c_int;
+    fn pidfile_remove(pfh: *mut pidfh) -> libc::c_int;
+}
+
+/// An open PID file.
+///
+/// A process that manages to construct this type holds an exclusive lock on the
+/// PID file.
+///
+/// Dropping the type will attempt to call [`remove`](PidFile::remove), but any
+/// errors will be suppressed. Call `remove` manually if you need to handle PID
+/// file removal errors.
+pub struct PidFile(*mut pidfh);
+
+impl PidFile {
+    /// Attempts to open and lock the specified PID file.
+    ///
+    /// If the file is already locked by another process, it returns
+    /// `Error::AlreadyRunning`.
+    pub fn open<P>(path: P) -> Result<PidFile, Error>
+    where
+        P: AsRef<Path>,
+    {
+        PidFile::open_with(path, Permissions::from_mode(0o600))
+    }
+
+    /// Like [`open`](PidFile::open), but opens the file with the specified
+    /// permissions rather than 0600.
+    pub fn open_with<P>(path: P, permissions: Permissions) -> Result<PidFile, Error>
+    where
+        P: AsRef<Path>,
+    {
+        let path_cstring = CString::new(path.as_ref().as_os_str().as_bytes())?;
+        let mut old_pid: libc::pid_t = -1;
+        let mode: libc::mode_t = permissions
+            .mode()
+            .try_into()
+            .expect("file permissions not valid libc::mode_t");
+        let f = unsafe { pidfile_open(path_cstring.as_ptr(), mode, &mut old_pid) };
+        if !f.is_null() {
+            let r = unsafe { pidfile_write(f) };
+            if r == 0 {
+                Ok(PidFile(f))
+            } else {
+                Err(Error::Io(io::Error::last_os_error()))
+            }
+        } else {
+            let err = io::Error::last_os_error();
+            if err.kind() == io::ErrorKind::AlreadyExists {
+                Err(Error::AlreadyRunning {
+                    pid: (old_pid != -1).then(|| old_pid),
+                })
+            } else {
+                Err(Error::Io(err))
+            }
+        }
+    }
+
+    /// Closes the PID file and removes it from the filesystem.
+    pub fn remove(mut self) -> Result<(), Error> {
+        let r = unsafe { pidfile_remove(self.0) };
+        self.0 = ptr::null_mut();
+        if r == 0 {
+            Ok(())
+        } else {
+            Err(Error::Io(io::Error::last_os_error()))
+        }
+    }
+}
+
+impl Drop for PidFile {
+    fn drop(&mut self) {
+        if self.0 != ptr::null_mut() {
```

Comment:
I'd usually expect this to call self.remove and throw the error away, does something prevent that?

Reply:
Remove takes a `mut self`, which consumes it, while Drop impls only get an `&mut self` borrow. One workaround is to have a private `&mut self` that they both call.

I almost suggested that this might have an issue with Drop getting called at the end of remove, but managed to notice that you've already handled that with the `self.0 != ptr::null_mut()`. May I suggest either a comment or tossing the inside bit of PidFile into an Option, which you `if let Some(...) = self.0.take() {` on remove/drop?

Reply:
Yep, added a comment.